keras/utils/training_utils.py

```diff
@@ -19,6 +18,35 @@ def _normalize_device_name(name):
     return name
 
 
+class DeviceSlice(Layer):
+    def __init__(self, index, total_devices, **kwargs):
+        self.index = index
+        self.total_devices = total_devices
+        super(DeviceSlice, self).__init__(**kwargs)
+
+    def call(self, inputs):
+        shape = K.tf.shape(inputs)
```

Comment:
Why K.tf? We should always use Keras backend instead of native backends.

Reply:
Well, a couple of things. First of all, the code this is used by is already marked as Tensorflow only. See [training_utils.py line 101-103](https://github.com/keras-team/keras/pull/9330/commits/415685102fd80ecd04dc8799ca55be67c0e4a287#diff-c18336769f2be165aae86963133c3a1fL100). I get why you're asking, though, since this layer could be used with arbitrary backends. However, I'm basically just rearranging existing code to make Keras not crash when doing ModelCheckpoints on multi-GPU workloads. I don't know the appropriate backend-agnostic functions for this If there's a better way to do it, then I'm all for it.

Reply:
Right so a good way to move forward would be to translate all the `tf` backend part to `K` backend and if you see that an operator is not available we can evaluate the situation. You just have to search for the mapping in the keras backend. I think that most of the operators you use are available if not all.